### PR TITLE
Improving spinlock pools

### PR DESCRIPTION
**Pull request by sithhell**
This patch introduces a multiplicative hashing algorithm for the spinlock pools to avoid modulo operations in the critical paths.
This, together with #3498 shows up to 10% performance increase on the Phylanx simple loop benchmark: https://docs.google.com/spreadsheets/d/15aNdbbyrWg3L56kY5Ud56sflX3Iy3tqbOC5dBaDijsc/edit#gid=0

### CODE REVIEW

**Comment by msimberg** [comment]
Does the size of the spinlock pool make much of a difference?

**Comment by hkaiser** [comment]
> Does the size of the spinlock pool make much of a difference?

Not sure. Using a prime number seems to be a good idea, though. Also, I'd suggest to pad the locks such as each is loaded into a separate cache line. That should help decreasing false sharing.

File: hpx/util/spinlock_pool.hpp [hunk 1]
```diff
@@ -41,20 +42,20 @@ namespace hpx { namespace util
     }
 #endif
 
-    template <typename Tag>
+    template <typename Tag, std::size_t N = 128>
     class spinlock_pool
```
[the remark this answers is not in the document]
Comment (sithhell) [reply hpx/util/spinlock_pool.hpp]: In theory, yes. In this implementation, no: The reason is the shift operation instead of division. We could manually clamp the value, but that would add another operation. I think requiring a power of two isn't that bad.

**Comment by sithhell** [comment]
> > Does the size of the spinlock pool make much of a difference?
> 
> Not sure. Using a prime number seems to be a good idea, though. Also, I'd suggest to pad the locks such that each is loaded into a separate cache line. That should help decreasing false sharing.

No, using a prime number instead won't change anything. In fact, it would eliminate the purpose of having a fast implementation of the hash. It is all explained nicely here: https://probablydance.com/2018/06/16/fibonacci-hashing-the-optimization-that-the-world-forgot-or-a-better-alternative-to-integer-modulo/ (also part of the fibhash header).

Increasing the size makes a difference for sure: The bigger the array, the less conflicts we have. I haven't played around with it though.

Padding the locks is definitely something we can try.

**Comment by hkaiser** [comment]
@sithhell, @msimberg ping?

**Comment by msimberg** [comment]
This would probably be good to go as is, but would it be worth it to make the default 128 (CMake) configurable? Or at least try a few other values for `N` before we merge this?

**Comment by sithhell** [comment]
I updated the PR to make the size configurable and pad to cache line boundaries.

The number not being a prime isn't a problem at all, since the fibhash itself takes care of the proper distribution. I tested with different sizes but couldn't observe significant performance difference (unless the number is too small) with future_overhead and phylanx simple_loop_test.

File: hpx/lcos/local/spinlock_pool.hpp [hunk 2]
```diff
@@ -14,39 +14,46 @@
...
+        template <typename Tag, std::size_t N>
         struct itt_spinlock_init
         {
             itt_spinlock_init();
             ~itt_spinlock_init();
         };
-    }
 #endif
+        // special struct to ensure cache line alignemnt
+        struct spinlock_holder
```
Comment (msimberg) [inline hpx/lcos/local/spinlock_pool.hpp]: Couldn't `spinlock` have `alignas(64)` instead?

Comment (sithhell) [reply hpx/lcos/local/spinlock_pool.hpp]: Yes, I wasn't sure if that was working on all our platforms... I remember that some had problems with over alignment

File: hpx/lcos/local/spinlock_pool.hpp [hunk 3]
```diff
@@ -14,39 +14,46 @@
...
+        {
+            // pad to 64 bytes
+            char cacheline_pad[64 - sizeof(lcos::local::spinlock)];
```
[the remark this answers is not in the document]
Comment (biddisco) [reply hpx/lcos/local/spinlock_pool.hpp]: I would suggest we add an `hpx::topology::get_cache_line_size` and one for `page_size` as well. then there is a single place where all code can get the value from and it can be adjusted by platform/architecture in a less invasive manner than having 64 and 4096 in code scattered about.

[the remark this answers is not in the document]
Comment (hkaiser) [reply hpx/lcos/local/spinlock_pool.hpp]: Good thinking!

**Comment by hkaiser** [comment]
@sithhell please verify my changes I applied in an attempt to address the review comments

File: hpx/lcos/local/spinlock.hpp [hunk 4]
```diff
@@ -15,6 +15,7 @@
 #include <hpx/config.hpp>
 
 #include <hpx/runtime/threads/thread_helpers.hpp>
+#include <hpx/runtime/threads/topology.hpp>
```
Comment (sithhell) [inline hpx/lcos/local/spinlock.hpp]: why did you add this include?

Comment (hkaiser) [reply hpx/lcos/local/spinlock.hpp]: Nice catch - it's a leftover. Will remove.

Comment (hkaiser) [reply hpx/lcos/local/spinlock.hpp]: Done.
